components/core/src/crypto/keys/cache.rs

```diff
@@ -0,0 +1,515 @@
+use crate::{crypto::{hash,
+                     keys::{encryption::{generate_origin_encryption_key_pair,
+                                         generate_service_encryption_key_pair,
+                                         generate_user_encryption_key_pair},
+                            generate_signing_key_pair,
+                            KeyFile,
+                            NamedRevision,
+                            OriginPublicEncryptionKey,
+                            OriginSecretEncryptionKey,
+                            PublicOriginSigningKey,
+                            RingKey,
+                            SecretOriginSigningKey,
+                            ServicePublicEncryptionKey,
+                            ServiceSecretEncryptionKey,
+                            UserPublicEncryptionKey,
+                            UserSecretEncryptionKey}},
+            error::{Error,
+                    Result},
+            fs::AtomicWriter};
+use std::{convert::TryFrom,
+          io::Write,
+          path::{Path,
+                 PathBuf}};
+
+/// Represents the location of all Habitat keys (user, service,
+/// origin, signing, and ring) locally on disk, as well as the APIs
+/// for retrieving and storing keys.
+#[derive(Clone, Debug, PartialEq)]
+pub struct KeyCache(PathBuf);
+
+impl AsRef<Path> for KeyCache {
+    /// Expose the path to this key cache.
+    fn as_ref(&self) -> &Path { self.0.as_ref() }
+}
+
+impl KeyCache {
+    pub fn new<P>(path: P) -> Self
+        where P: Into<PathBuf>
+    {
+        KeyCache(path.into())
+    }
+
+    /// Ensure that the directory backing the cache exists on disk.
+    pub fn setup(&self) -> Result<()> {
+        if !self.0.is_dir() {
+            std::fs::create_dir_all(&self.0)?;
+        }
+        Ok(())
+    }
```

Comment:
Maybe... I'd like that conceptually, at any rate. There may still be a few places where it would be awkward to have a `Result`, rather than in infallible operation, but we could probably weed those out.

This might be another follow-on PR.